RIOT-OS/RIOT
# Hcsr04 module

### Contribution description

Hello my name in Evripidis and i created a driver module for the hcsr04 ultrasonic distance sensor, in order to familiarize myself with RIOT and be better prepared for my diploma.

I have tried to follow and the guideline steps mentioned in the documentation page and document each function.

This module contains all the basic driver structure and also exposes its functions to the saul API. It also contains a temperature setting so as to improve the distance estimation due to change in the air speed.

It also contains some error detection should the maximum read frequency is surpassed, or when waiting for a echo response pulse and a timeout occurs, maybe due to connection issues.





### Testing procedure

I have successfully tested the driver on an esp32 & esp8266 mcu's. It seems to work all right. To test it, just run the main.c file in the test/hcsr04 directory. The test outputs 
3 series of measurements:
- firstly with trigger/write functions
- secondly using the get_distance
- lastly using the saul API

Note that, in order not to avoid conflict with the saul registry, i deactivate it in the beginning of the test, and later i manually set it up again.
Also should any error condition is met in this test, the kernel will crush due to the use of the assert function. So if you want to observe the error codes from the driver module, the assert functions must be removed.

While it is a pretty simple sensor, it was still fun developing this driver.
If you have any comment or you would like me to make any changes, pls let me know.
Also if you have comments on the code quality or styling it will be very helpful. 



### Issues/PRs references
No known issues yet.
